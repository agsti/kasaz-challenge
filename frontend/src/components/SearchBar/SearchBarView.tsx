import React, {useState} from "react";
import SearchBox from "./SearchBox";

import ArrowIcon from "./icons/arrow-icon.svg";

import "./SearchBar.css";

type SearchBarProps = {
    onSearch: (arg0: string) => void,
    searchValue: string,
    setSearchValue: (arg0: string) => void,
    searchHasChanged: boolean,
    onFilterButtonClicked: ()=>void,
    filterArrowLookingUp: boolean
}
export default function SearchBarView(props:SearchBarProps) {
    const {
        onSearch,
        setSearchValue, 
        searchValue,
        searchHasChanged,
        onFilterButtonClicked, 
        filterArrowLookingUp
    } = props

    return (
        <div className="search-container">
            <SearchBox
                onChange={(e) => {setSearchValue(e.target.value)}}
                value={searchValue}
                onSearch={onSearch}
                showSearchButton={searchHasChanged}
            />
            <div className="filter-btn" onClick={()=> onFilterButtonClicked()}>
                Filtros
                <img className={filterArrowLookingUp? "filter-icon filter-menu-shown" : "filter-icon"} src={ArrowIcon} />
            </div>
        </div>
    )
}
