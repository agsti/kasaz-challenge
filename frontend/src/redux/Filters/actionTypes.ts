export const SET_MIN_PRICE = "SET_MIN_PRICE";
export const SET_MAX_PRICE = "SET_MAX_PRICE";
export const SET_MIN_SIZE = "SET_MIN_SIZE";
export const SET_MAX_SIZE = "SET_MAX_SIZE";
export const SET_MIN_ROOMS = "SET_MIN_ROOMS";
export const SET_CITY = "SET_CITY";
export const SET_FILTERS_OUT_OF_SYNC="SET_FILTERS_OUT_OF_SYNC";
