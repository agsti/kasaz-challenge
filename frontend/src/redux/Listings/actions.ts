import { Listing } from "./reducer";
import { SET_SCROLL_POSITION, ADD_LISTINGS, SET_LISTINGS } from "./actionTypes";

export const AddListings = (listings : Listing[]) => ({type: ADD_LISTINGS, payload: listings})

export const SetListings = (listings : Listing[]) => ({type: SET_LISTINGS, payload: listings})

export const SetScrollPosition = (scroll : number) => ({type: SET_SCROLL_POSITION, payload: scroll})