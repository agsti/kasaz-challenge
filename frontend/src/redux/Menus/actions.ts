import {TOGGLE_FILTER_MENU} from "./actionTypes";


export const ToggleFilterMenu = () => ({type: TOGGLE_FILTER_MENU})
