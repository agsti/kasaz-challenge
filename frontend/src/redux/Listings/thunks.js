import { GetListings } from "../../api";
import { SetListings } from "./actions";


const mapApiToFE = (apiReponse) =>{
    return {
        title: apiReponse.Title,
        price: apiReponse.Price,
        priceSqm: 3.21,
        sqm: apiReponse.AreaSqm,
        nRooms: apiReponse.NRooms,
        nBath: apiReponse.NBathrooms,
        picture: apiReponse.ImageUrl
    }
}


export const GetNewListings = () => {
    return function(dispatch , getState) {
        const {filters} = getState()
 
        return GetListings(
            filters.minSize, 
            filters.maxSize,
            filters.minPrice, 
            filters.maxPrice, 
            filters.minRooms, 
            0
        ).then(r =>{
            dispatch(SetListings(r.data.map(mapApiToFE)))
        })
    }
}