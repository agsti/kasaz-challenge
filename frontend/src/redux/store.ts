import { createStore, combineReducers, applyMiddleware} from 'redux';

import thunk from 'redux-thunk';

import {FiltersReducer, FiltersInitialState} from "./Filters/reducer";
import {MenuReducer, MenusInitialState} from "./Menus/reducer";

import {ListingsReducer, ListingsInitialState} from "./Listings/reducer";

const InitialState = {
    filters: FiltersInitialState,
    menus: MenusInitialState,
    listings: ListingsInitialState
}

const reducers = combineReducers({
    filters: FiltersReducer,
    menus: MenuReducer,
    listings: ListingsReducer
})

const store = createStore(reducers, applyMiddleware(thunk))

export type RootStateType = typeof InitialState;

export default store;
