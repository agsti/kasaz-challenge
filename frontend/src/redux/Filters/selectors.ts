import {RootStateType} from "../store";

export const SelectFilterState = (state: RootStateType) => state.filters
