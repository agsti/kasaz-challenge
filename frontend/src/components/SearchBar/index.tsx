import React from "react";
import SearchBarView from "./SearchBarView";
import {useSelector, useDispatch} from "react-redux";

import {SelectMenuState} from "../../redux/Menus/selectors";
import {SetCity, SetFilterOufOfSync} from "../../redux/Filters/actions";
import {ToggleFilterMenu} from "../../redux/Menus/actions";
import { GetNewListings } from "../../redux/Listings/thunks";
import { SelectFilterState } from "../../redux/Filters/selectors";


export default function SearchBar() {
    const menuState = useSelector(SelectMenuState);
    const filterState = useSelector(SelectFilterState);
    const dispatch = useDispatch();

    return (
        <SearchBarView
            onSearch={search => {
                dispatch(GetNewListings())
                dispatch(SetFilterOufOfSync(false))
                dispatch(ToggleFilterMenu())
            }}
            onFilterButtonClicked={ () =>{
                dispatch(ToggleFilterMenu())
            }}
            filterArrowLookingUp={menuState.showFilterMenu}
            searchHasChanged={filterState.outOfSync}
            searchValue={filterState.city}
            setSearchValue={v => {
                dispatch(SetCity(v))
                !filterState.outOfSync && dispatch(SetFilterOufOfSync(true))
            }
            }
    />
    )
}
