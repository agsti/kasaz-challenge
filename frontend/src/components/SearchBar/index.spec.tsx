import React from 'react';
import 'jsdom-global/register';

import * as ReactRedux from "react-redux";
import thunk from "redux-thunk";

import { mount, ReactWrapper } from 'enzyme';
import configureStore from 'redux-mock-store';
import { describe, expect, it } from '@jest/globals';




import { Provider } from 'react-redux';

import { RootStateType } from '../../redux/store';

import SearchBar from './index';
import SearchBarView from './SearchBarView';

jest.mock("../../redux/Listings/thunks", ()=>({
    GetNewListings: jest.fn(()=>({type: "TYPE"}))
}))
jest.mock("../../redux/Filters/actions")
jest.mock("../../redux/Menus/actions")



import { GetNewListings } from "../../redux/Listings/thunks";
import { SetFilterOutOfSync } from "../../redux/Filters/actions";
import { ToggleFilterMenu } from "../../redux/Menus/actions";




describe('FilterMenu', () => {
    let store;
    let component : ReactWrapper;
    let mockStore;

    const state : Partial<RootStateType> = {
        filters: {
            minPrice: 10,
            maxPrice: 20,
            city: "city",
            minSize: 30,
            maxSize: 40,
            minRooms:3,
            outOfSync: true
        },
        menus: {
            showFilterMenu: true
        }
    }

    beforeEach(() => {
        mockStore = configureStore([thunk]);

        store = mockStore(state);
        
        component = mount(
            <Provider store={store}>
                <SearchBar />
            </Provider>
        );
    });


    it('should render correctly', () => {
        expect(component).toMatchSnapshot()
    })

    it('should pass correct properties to view', () =>{
        const view = component.find(SearchBarView)
        
        expect(view.prop("filterArrowLookingUp")).toBe(state.menus.showFilterMenu)
        expect(view.prop("searchHasChanged")).toBe(state.filters.outOfSync)
        expect(view.prop("searchValue")).toBe(state.filters.city)

    })

    // it('onSearch should dispatch correct actions', () =>{
  

    //     const mockDispatch = jest.fn()
    //     jest.mock("react-redux", ()=>({
    //         useDispatch: mockDispatch
    //     }))

    //     const view = component.find(SearchBarView)
        
    //     const ret = view.invoke('onSearch')("search-data")
    //     expect(GetNewListings).toBeCalledTimes(1)
    // })
})
