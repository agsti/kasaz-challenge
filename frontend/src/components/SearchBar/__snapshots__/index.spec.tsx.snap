// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterMenu should render correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <SearchBar>
    <SearchBarView
      filterArrowLookingUp={true}
      onFilterButtonClicked={[Function]}
      onSearch={[Function]}
      searchHasChanged={true}
      searchValue="city"
      setSearchValue={[Function]}
    >
      <div
        className="search-container"
      >
        <SearchBox
          onChange={[Function]}
          onSearch={[Function]}
          showSearchButton={true}
          value="city"
        >
          <span
            className="search-box"
          >
            <input
              className="search-input"
              onChange={[Function]}
              value="city"
            />
            <span
              className="search-icon"
            >
              <img
                onClick={[Function]}
                src=""
              />
            </span>
          </span>
        </SearchBox>
        <div
          className="filter-btn"
          onClick={[Function]}
        >
          Filtros
          <img
            className="filter-icon filter-menu-shown"
            src=""
          />
        </div>
      </div>
    </SearchBarView>
  </SearchBar>
</Provider>
`;
